### ISSUE TITLE
Enhancing the events.tsv file for MEG, EEG, iEEG with 'channels' and 'how-it-was-marked'

### ISSUE THREAD

**Issue by adam2392** [bids-standard/bids-specification#959]
In BIDS-EEG Derivatives discussion, there's some discussion on difference between annotations.tsv and events.tsv files. Tbh not sure where we've converged there, but many ppl feel like there might be some redundancy there, so annotations.tsv might not be needed.

### Enhancement Proposal
I think this was brought up before, but is it possible to add a `channels` column that specifies an OPTIONAL list of channels that the event occurred in (e.g. a bad marking). 

Moreover, to differentiate the events.tsv file markings from manual and automated markings, there would need to be an extra RECOMMENDED column of whether or not the marking was detected visually, or with an automated algorithm. For example, `visual` vs `RMSalgorithm` for e.g. a high-frequency oscillation marking on a dataset.

This would enable us tools that have events-with-channel-knowledge to have a complete roundtrip (e.g. in MNE-BIDS) with events.tsv because they can have channels now that is documented in BIDS.

cc: @hoechenberger  @guiomar @sappelhoff @robertoostenveld (tag others who I missed?)

**Comment by VisLab** [comment]
I think this case is already handled in HED.

HED-annotated data would generally include a tag from the `Event` tree such as `Sensory-event` or `Data-feature`.  If you decide to have any column in your particular events file designate the "event type", you just need to define what those types mean in terms of HED tags in a top-level JSON sidecar. 

Maybe we can talk off line about how MNE-BIDS might interact with HED?

**Comment by adam2392** [comment]
From my understanding HED handles what are the contents (`description`) of each event.

Re specifying channels of events: Looking at https://bids-specification.readthedocs.io/en/stable/99-appendices/03-hed.html, I don't see where one would annotate events that occur in time but only for a specific set of channels.

Re specifying manual/automated marking of events: Is this where HED would come in? Would you be able to elaborate how that might look? I'm not super familiar w/ HED.

**Comment by VisLab** [comment]
The bids specification for HED needs to be updated to accommodate the new version of HED (released Aug. 2021), which is more capable.  The bids validator now supports most but not all of the features in the new version of HED.

Here is a link to the documentation: [https://hed-specification.readthedocs.io/en/latest/index.html](https://hed-specification.readthedocs.io/en/latest/index.html).  

Here is a link to a [paper](https://www.sciencedirect.com/science/article/pii/S1053811921010387) in press that gives a detailed BIDS example.

If you decided to designate a column in the events file to mark a single channel, you could
put the channel number in there and use the value substitution for HED very easily.  You could also you the HED column in the events file to explicitly tag certain events.

However, it seems as though you are thinking of putting a list of channels as an a single entry in the events file.  Right now HED only handles single values in columns.  I am not sure what is allowed in BIDS as far as lists of values. I thought they had to be single numbers or possibly a string representing a list.

**Comment by adam2392** [comment]
Ultimately, I would like something like this:

| trial_type  | ... | channel  | manual |
| ------------- | ------------- | ------------- | ------------- |
| event_one | ...  | C1  | False |
| event_one  | A12  | False |

which shows that `event_one` was identified in `C1`, `A12` channels and using some automated algorithm.

For HED, it seems you would need something along these lines for the *events.tsv file:

| trial_type  | ... | channel  | manual |
| ------------- | ------------- | ------------- | ------------- |
| event_one | ...  | subset_01  | False |

with an accompanying json:

```
{
   "channel": {
       "Description": "Channels at which event occurs, or some other definition the user wants for channels",
       "Levels": {
          "subset_01": ["C1", "A12"],
         "other possible subsets of channels",
       },
   }
}
```

This approach to marking which set of channels are part of a specific event seems pretty verbose and the HED description of channel is then subjective for the user. 

Assuming the use case of marking an event specifically for a channel is common: wouldn't it be better to allow append an OPTIONAL `channel` column in BIDS spec for `events.tsv` file that describes what it can be?

**Comment by hoechenberger** [comment]
I agree that having a `channel` column in `events.tsv` seems to be a pragmatic solution here. Maybe it should even be called `channels` and expect an array of channel names, even if only specifying a single channel? How would we label "all channels"?

**Comment by VisLab** [comment]
Hmm....It looks like you are specifying a separate row in the event file for each channel?  This is what HED would like also.

The JSON above doesn't have any HED tags at all.  The description and levels are ignored by HED. The HED part of a JSON side car might be:

```json
{
    "channel_name":  {
          "HED": "Channel/#"
     }
    "annotate_type": {
          "HED": "Label/#"
     }
}
```
This assumes that we put in a HED tag called `Channel`.  Otherwise we would need to define one and use `Def/Channel/#` instead of `Channel/#`.   The # is replaced by actual column value when the annotations are assembled.

The table:

onset | trial_type | ... | channel_name |annotation_type
------ | ---------- | -- | --------------- | --
10.32 | event_one | ... | C1 | Artifact
10.32 | event_one | ... | A12 | Artifact

corresponds to assembled events: 

onset  | HED
------- | --
10.32  |  "Channel/C1, Label/Artifact"
10.32  |  "Channel/A12, Label/Artifact"

If other columns had entries in the JSON sidecar additional annotations would be included.

Note:  Users are free to add any columns they want to event files, so there is nothing to stop them from adding a `channel` or a `channels` column. However, I think caution should be used when added recommended or required columns for particular modalities.  ANY column can  already be added optionally.

**Comment by robertoostenveld** [comment]
The current BIDS specification already allows you to add a `channel` column to the `events.tsv`, and to document your particular use for that in the  `events.json`. The only columns that are REQUIRED are `onset` and `duration`.  

If you go to the original link and scroll down a bit, you see 
> An arbitrary number of additional columns can be added. Those allow describing other properties of events that could be later referred in modelling and hypothesis extensions of BIDS. Note that the trial_type and any additional columns in a TSV file SHOULD be documented in an accompanying JSON sidecar file.

So in that respect no change to BIDS is needed. If you want to standardize how the values in the `channel` column are to be stored and interpreted (as we have for `sample`, `trial_type`, `response_time`, `value` and `HED`), then that would need to be an extension to BIDS. E.g. that precise specification should include how channels are coded (numerical zero-offset or one-offset, or as labels corresponding to `channels.tsv`) and in which cases channels apply (which seems obvious for EEG, but consider an `events.tsv` that accompanies a `func.nii.gz`, or what if there is a `physio.tsv`?)

**Comment by adam2392** [comment]
> So in that respect no change to BIDS is needed. If you want to standardize how the values in the `channel` column are to be stored and interpreted (as we have for `sample`, `trial_type`, `response_time`, `value` and `HED`), then that would need to be an extension to BIDS. E.g. that precise specification should include how channels are coded (numerical zero-offset or one-offset, or as labels corresponding to `channels.tsv`) and in which cases channels apply (which seems obvious for EEG, but consider an `events.tsv` that accompanies a `func.nii.gz`, or what if there is a `physio.tsv`?)

The proposal here is specifically as you say to specify how exactly 'channels' column is used. I agree that there is always the possibility of using HED or say custom columns described with a sidecar JSON, but this makes downstream validation and tooling more difficult because it means more then one organization is valid. 

Since channels is a common component of certain task events for MEG, EEG and iEEG, then I was hoping to clarify exactly how it should look and be stored. That way the validator and downstream tools can consolidate their expectations of the input and output format. I wouldnt specify this for other data modalities as it doesn't make sense.

Do you think this is valid for a PR to the task events page specifically for the MEG, EEG and iEEG modalities?

**Comment by robertoostenveld** [comment]
> Do you think this is valid for a PR to the task events page specifically for the MEG, EEG and iEEG modalities?

I think it is better to start with explicit examples that can be discussed (this can be in actually shared datasets), and a prototype implementation in the tooling as you say. Once those receive positive responses from the community, a PR to the standard would be appropriate. Doing it the other way around would result in it remaining rather theoretical, making the required community support and consensus for the PR harder to achieve and the overall process less efficient.

**Comment by robertoostenveld** [comment]
Perhaps it is goot to tag the whole @bids-standard/bep021 on this.

**Comment by CPernet** [comment]
+1 with @robertoostenveld working on an example trying with existing capabilities is key -- if from example one can demonstrate this doesn't work then an extension/change makes sense
